src/metabase/lib/js.cljs

```diff
@@ -339,6 +339,41 @@
                                     ;; match up. Therefore de-dupe with `frequencies` rather than simply `set`.
                                     (assoc inner-query :fields (frequencies fields)))))))
 
+(defn- compare-legacy-field-refs
+  [[key1 id1 opts1]
+   [key2 id2 opts2]]
+  ;; A mismatch of `:base-type` or `:effective-type` when both x and y have values for it is a failure.
+  ;; If either ref does not have the `:base-type` or `:effective-type` set, that key is ignored.
+  (letfn [(clean-opts [o1 o2]
+            (not-empty
+              (cond-> o1
+                (not (:base-type o2))      (dissoc :base-type)
+                (not (:effective-type o2)) (dissoc :effective-type))))]
+    (= [key1 id1 (clean-opts opts1 opts2)]
+       [key2 id2 (clean-opts opts2 opts1)])))
+
+(defn- query=* [x y]
+  (cond
+    (and (vector? x)
+         (vector? y)
+         (= (first x) (first y) :field))
+    (compare-legacy-field-refs x y)
+
+    ;; Otherwise this is a duplicate of clojure.core/=.
+    (and (map? x) (map? y))
+    (and (= (set (keys x)) (set (keys y)))
+         (every? (fn [[k v]]
+                   (query=* v (get y k)))
+                 x))
+
+    (and (sequential? x) (sequential? y))
+    (and (= (count x) (count y))
+         (every? identity (map query=* x y)))
+
+    ;; Either mismatched map/sequence/nil/etc., or primitives like strings.
```

Comment:
Even if we don't currently have field refs in sets, I would feel better if we recursed into sets too.

Reply:
It's a pain to compare sets in this way. Since these are legacy queries and we don't use sets in them anywhere so far as I know, I've just added code to throw if sets are encountered.

Reply:
Actually I tried to write a test for that and it doesn't really work. This is consuming legacy queries, and `js->clj` doesn't convert JS `Set`s into CLJS sets (it treats them as opaque JS objects).

Reply:
Maybe just a comment about the need to revisit this if we ever want to use it for non-legacy queries.